### PR TITLE
refactor change code, add TextRange and TextSize types

### PR DESCRIPTION
**Pull request by oberblastmeister**
I have put the new `Change` in a separate file and haven't updated everything yet.

### CODE REVIEW

**Comment by kirawi** [comment]
I'm confused, was there prior discussion around this on Matrix? I don't quite understand the reason for this change.

**Comment by cessen** [comment]
I really appreciate your enthusiasm to contribute to Helix.  However, it's not at all clear to me what the benefit of this refactor is, nor the new types you're adding.  It feels like abstraction for abstraction's sake, and to my eye at least actually makes things less straight-forward/obvious.

Maybe I missed where this was discussed, but I'd really recommend opening an issue to discuss sweeping changes like this and get some consensus before actually coding them up.  That way you don't spend a lot of effort that just gets rejected.  (Again, maybe I just missed where that discussion happened, in which case nevermind.)

**Comment by oberblastmeister** [comment]
I did talk on matrix about something similar but that was just for me to understand things. There was no prior discussion matrix about this refactor, The reason for this change is that this greatly simplifies the code. A single `Change` can now be taken in isolation. There are no unnecessary conversions between changes and `Operation` and then back. Changing to `Operation` doesn't even help because it makes processing even harder as it cannot be taken in isolation which means processing them must repeat the same mutable code.  This code also enforces stronger invariants. It makes sure that every Change is disjoint and that they are sorted. Not doing this can cause issues. In addition, this pr adds `TextSize` and `TextRange` which are more ergonomic types. They inforce more invariants than representing them as tuples and have multiple useful functions that I have seen repeated in code.

**Comment by oberblastmeister** [comment]
Thanks for appreciating my enthusiasm! I should definitely post an issue next time, sorry. I think I definitely am not overabstracting though. I removed the `Operation` abstraction, which I thought was unnecessary for the reasons above.`TextSize` and `TextRange` are nice abstractions, especially when we are in an editor that uses these sort of things all the time.  The stuff in the `one` module I am not sure about yet. I think they are nice because they allow types to be 'correct by construction'.

**Comment by oberblastmeister** [comment]
@cessen Thanks for your suggestions. I have simplified the code a lot, it is now really nice. I think I did overabstract the `Change` struct, so now it is much better. Please give this another chance!

File: helix-core/src/lib.rs [hunk 1]
```diff
@@ -110,12 +112,12 @@ pub use {regex, tree_sitter};
...
+pub use selection::{SelectionRange, Selections};
 pub use smallvec::SmallVec;
 pub use syntax::Syntax;
 
 pub use diagnostic::Diagnostic;
 pub use state::State;
 
 pub use line_ending::{LineEnding, DEFAULT_LINE_ENDING};
-pub use transaction::{Assoc, Change, ChangeSet, Operation, Transaction};
+pub use transaction::{Assoc, Change, ChangeSet, Operation, Transaction};
```
Comment (kirawi) [inline helix-core/src/lib.rs:123]: Missing newline here.

File: helix-core/src/text_size/offset.rs [hunk 2]
```diff
@@ -0,0 +1,77 @@
+use std::{convert::TryFrom, num::TryFromIntError, ops::{Add, AddAssign, Sub, SubAssign}};
```
Comment (kirawi) [inline helix-core/src/text_size/offset.rs:1]: Sorry, I'm confused on the purpose of this file.

File: helix-core/src/text_size/range.rs [hunk 3]
```diff
@@ -0,0 +1,400 @@
...
+    ops::{self, Add, Sub},
+};
+
+use super::size::TextSize;
+
+/// A range in text, represented as a pair of [`TextSize`][struct@TextSize].
+///
+/// It is a logic error for `start` to be greater than `end`.
+#[derive(Default, Copy, Clone, Eq, PartialEq, Hash)]
+pub struct TextRange {
```
Comment (kirawi) [inline helix-core/src/text_size/range.rs:13]: Isn't this basically just `SelectionRange`?

File: helix-core/src/text_size/traits.rs [hunk 4]
```diff
@@ -0,0 +1,25 @@
+use super::size::TextSize;
```
Comment (kirawi) [inline helix-core/src/text_size/traits.rs:1]: I'm also confused here, I feel like this is also already a covered case w/ `RopeGraphemes` (I might be wrong though).

File: helix-core/src/transaction.rs [hunk 5]
```diff
@@ -1,7 +1,11 @@
-use crate::{Range, Rope, Selection, Tendril};
+use crate::{SelectionRange, Rope, Selections, Tendril, text_size::TextRange};
 use std::borrow::Cow;
 
 /// (from, to, replacement)
+// pub struct Change {
```
Comment (kirawi) [inline helix-core/src/transaction.rs:5]: Dead code.

File: helix-core/src/text_size/traits.rs [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+        (self.len_utf8() as u32).into()
+    }
+}
+
+/// identity
+impl TextLen for TextSize {
+    fn text_len(&self) -> TextSize {
+        *self 
+    }
+}
```
Comment (kirawi) [inline helix-core/src/text_size/traits.rs:25]: Missing newline.

File: helix-core/src/text_size/size.rs [hunk 7]
```diff
@@ -0,0 +1,189 @@
...
+
+impl<A> iter::Sum<A> for TextSize
+where
+    TextSize: Add<A, Output = TextSize>,
+{
+    #[inline]
+    fn sum<I: Iterator<Item = A>>(iter: I) -> TextSize {
+        iter.fold(0.into(), Add::add)
+    }
+}
```
Comment (kirawi) [inline helix-core/src/text_size/size.rs:189]: Missing newline.

File: helix-core/src/transaction.rs [hunk 8]
```diff
@@ -1,7 +1,11 @@
...
+//     insert: Tendril,
+//     delete: TextRange,
+// }
 pub type Change = (usize, usize, Option<Tendril>);
```
Comment (kirawi) [inline helix-core/src/transaction.rs:9]: This conflicts with `struct Change`.

**Review by kirawi — COMMENTED** [review]
It's still not overtly clear to me what benefits this has over the existing code, but I'm not really familiar with this area of the codebase either. (I would also suggest marking this PR as a draft; there should be an option on the right-hand sidebar of this page)

File: helix-core/src/text_size/range.rs (hunk 3 above)
Comment (oberblastmeister) [reply helix-core/src/text_size/range.rs:13]: `SelectionRange` can be flipped

**Comment by oberblastmeister** [comment]
Could you explain what you still don't think has benefits over the existing code? I think I can keep `TextRange` and remove `TextSize` and `TextOffset` as those don't seem to be very useful. I think we can just use a type alias instead for `TextSize`.

**Comment by kirawi** [comment]
From the cursory look I gave, it seems to me that you can achieve the same benefit for less abstraction/work by improving upon the already-existing code in `transaction.rs` rather than pulling in all this external code. If you can flatten the `Change` processing, I think that would helpful. Again, I'm not really familiar with this area of the codebase (I only worked with it for `helix-core/src/diff.rs`).

**Comment by oberblastmeister** [comment]
Like I explained above, the existing `Change` code is much larger with more abstractions and isn't as safe as this code. It has to convert from and to `Operation` which is a worse representation in my opinion. For example, look at the function `invert`, in the new code it is like 2 lines of code. The existing implementation takes 20 something lines. This is similar when comparing many functions. Additionaly, the new code has the option to ensure that everything is disjoint and sorts the changes. This guarantees that everything happens properly, it is even listed as a todo in the code by blaz. Thats why I created this pr, to refactor the code and make it simpler. As for the other types I added, I think we can remove `TextSize` and `TextOffset` because those don't make much sense. I like `TextRange` however because it makes stuff nicer to work with. It also ensures that you cannot subtract with overflow which is currently not prevented. However it is not absolutely necessary and I can remove it if people really want me to.

**Comment by archseer** [comment]
I agree with @cessen's sentiment: I appreciate the effort put into this, but a sweeping change that affects _all_ open PRs needs a prior discussion (see https://github.com/helix-editor/helix/issues/362 for an example). In particular, it's good to talk through **why** the code is currently the way it is, because there's often a reason behind it.

> The reason for this change is that this greatly simplifies the code.

The diff is currently +1200 -500. It's also very noisy, most of the diff is renaming various existing structs, I'd prefer to keep those names as is.

> There are no unnecessary conversions between changes and `Operation` and then back. Changing to `Operation` doesn't even help because it makes processing even harder as it cannot be taken in isolation which means processing them must repeat the same mutable code.

There are of course reasons why we have two types right now. The editing primitive is an `Operation` which models an [operational transform](https://en.wikipedia.org/wiki/Operational_transformation) and it's map and compose (`map` currently being unimplemented) allow for far greater flexibility in the future (async formatting after some user edits? map the formatting transaction over the edits and apply it. It's also the foundation for collaborative editing).

But `Operation`s aren't very ergonomic, so the `change/change_by_range` functions expose an API that's easier to use. I did attempt to merge the two before, but it either made the OT operations more convoluted, or it made the API less easy to use. I would be open to such a change if we can come up with a proposal that's acceptable.

> Additionaly, the new code has the option to ensure that everything is disjoint and sorts the changes. This guarantees that everything happens properly, it is even listed as a todo in the code by blaz.

Since the changeset builder is more low-level and we control all the methods I was thinking there's no need to do a check for ordering and this should actually be a debug_assert. Kakoune has built-in assertions in debug builds that verify a bunch of invariants before and after every command.

**Pull request closed without merging**